Src/Couchbase.Linq.UnitTests/BucketContextTests.cs

```diff
@@ -12,95 +15,334 @@ public class BucketContextTests
+            var ctx = new BucketContext(bucket.Object);
+
+            //act-assert
+            Assert.Throws<CouchbaseWriteException>(() => ctx.Remove(beer));
+        }
+
+        [Test]
+        public void Save_When_KeyAttribute_Is_Not_Defined_Throw_DocumentIdMissingException()
+        {
+            //arrange
+            var brewery = new Brewery();
+            var bucket = new Mock<IBucket>();
+            var result = new Mock<IOperationResult<Brewery>>();
+            result.Setup(x => x.Status).Returns(ResponseStatus.Success);
+            bucket.Setup(x => x.Upsert(It.IsAny<string>(), It.IsAny<Brewery>())).Returns(result.Object);
+            var ctx = new BucketContext(bucket.Object);
+
+            //act-assert
+            Assert.Throws<KeyAttributeMissingException>(() => ctx.Save(brewery));
+        }
+
+        [Test]
+        public void Remove_When_Write_Is_Succesful_Return_Success()
+        {
+            //arrange
+            var beer = new Beer();
+            var bucket = new Mock<IBucket>();
+            var result = new Mock<IOperationResult<Beer>>();
+            result.Setup(x => x.Status).Returns(ResponseStatus.Success);
+            result.Setup(x => x.Success).Returns(true);
+            bucket.Setup(x => x.Remove(It.IsAny<string>())).Returns(result.Object);
+            var ctx = new BucketContext(bucket.Object);
+
+            //act-assert
+            ctx.Remove(beer);
+        }
+
+        [Test]
+        public void Remove_When_DocId_Is_Not_Defined_Throw_DocumentIdMissingException()
+        {
+            //arrange
+            var brewery = new Brewery();
+            var bucket = new Mock<IBucket>();
+            var result = new Mock<IOperationResult<Brewery>>();
+            result.Setup(x => x.Status).Returns(ResponseStatus.Success);
+            bucket.Setup(x => x.Upsert(It.IsAny<string>(), It.IsAny<Brewery>())).Returns(result.Object);
+            var ctx = new BucketContext(bucket.Object);
+
+            //act-assert
+            Assert.Throws<KeyAttributeMissingException>(() => ctx.Remove(brewery));
+        }
+
+        [Test]
+        public void BeginChangeTracking_ChangeTrackingEnabled_Is_True()
+        {
+            //arrange
+            var bucket = new Mock<IBucket>();
+            bucket.Setup(x => x.Configuration).Returns(new ClientConfiguration().BucketConfigs.First().Value);
+            var ctx = new BucketContext(bucket.Object);
+
+            //act
+            ctx.BeginChangeTracking();
+
+            //assert
+            Assert.IsTrue(ctx.ChangeTrackingEnabled);
+        }
+
+        [Test]
+        public void EndChangeTracking_ChangeTrackingEnabled_Is_False()
+        {
+            //arrange
+            var bucket = new Mock<IBucket>();
+            bucket.Setup(x => x.Configuration).Returns(new ClientConfiguration().BucketConfigs.First().Value);
+
+            var ctx = new BucketContext(bucket.Object);
+            ctx.BeginChangeTracking();
```

Comment:
Jeff,

Do you think we need another unit test for calling BeginChangeTracking twice, followed by a single EndChangeTracking, to ensure that change tracking is still operational?